pkg/controller/servicebindingrequest/servicebinder.go

```diff
@@ -236,40 +241,102 @@ func (b *serviceBinder) onError(
+	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, &b.sbr.Status)
+	if err != nil {
+		return requeueOnConflict(err)
+	}
+
+	// appending finalizer, should be later removed upon resource deletion
+	sbr.SetFinalizers(append(removeStringSlice(b.sbr.GetFinalizers(), finalizer), finalizer))
+	if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+		return noRequeue(err)
+	}
+
+	b.logger.Info(applicationError.Error())
+
+	if errors.Is(applicationError, errApplicationNotFound) {
+		sbr.SetFinalizers(nil)
+		if _, err = b.updateServiceBindingRequest(sbr); err != nil {
+			return noRequeue(err)
+		}
+		return requeueOnNotFound(errApplicationNotFound, requeueAfter)
+	}
+
+	return done()
+}
+
 // bind configures binding between the Service Binding Request and its related objects.
 func (b *serviceBinder) bind() (reconcile.Result, error) {
-	sbrStatus := b.sbr.Status.DeepCopy()
-
 	b.logger.Info("Saving data on intermediary secret...")
 
 	secretObj, err := b.secret.createOrUpdate(b.envVars, b.sbr.AsOwnerReference())
 	if err != nil {
 		b.logger.Error(err, "On saving secret data..")
-		return b.onError(err, b.sbr, sbrStatus, nil)
+		return b.onError(err, b.sbr, &b.sbr.Status, nil)
 	}
-	sbrStatus.Secret = secretObj.GetName()
+	b.sbr.Status.Secret = secretObj.GetName()
 
+	if isApplicationSelectorEmpty(b.sbr.Spec.ApplicationSelector) {
+		return b.handleApplicationError(EmptyApplicationSelectorReason, errEmptyApplicationSelector)
+	}
 	updatedObjects, err := b.binder.bind()
 	if err != nil {
 		b.logger.Error(err, "On binding application.")
-		return b.onError(err, b.sbr, sbrStatus, updatedObjects)
+		if errors.Is(err, errApplicationNotFound) {
+			return b.handleApplicationError(ApplicationNotFoundReason, errApplicationNotFound)
+		}
+		return b.onError(err, b.sbr, &b.sbr.Status, nil)
 	}
-	b.setApplicationObjects(sbrStatus, updatedObjects)
+	b.setApplicationObjects(&b.sbr.Status, updatedObjects)
 
 	// annotating objects related to binding
 	namespacedName := types.NamespacedName{Namespace: b.sbr.GetNamespace(), Name: b.sbr.GetName()}
 	if err = setAndUpdateSBRAnnotations(b.dynClient, namespacedName, append(b.objects, secretObj)); err != nil {
 		b.logger.Error(err, "On setting annotations in related objects.")
-		return b.onError(err, b.sbr, sbrStatus, updatedObjects)
+		return b.onError(err, b.sbr, &b.sbr.Status, updatedObjects)
 	}
 
-	conditionsv1.SetStatusCondition(&sbrStatus.Conditions, conditionsv1.Condition{
-		Type:   BindingReady,
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:   CollectionReady,
+		Status: corev1.ConditionTrue,
+	})
+	conditionsv1.SetStatusCondition(&b.sbr.Status.Conditions, conditionsv1.Condition{
+		Type:   InjectionReady,
 		Status: corev1.ConditionTrue,
 	})
 
 	// updating status of request instance
-	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, sbrStatus)
+	sbr, err := b.updateStatusServiceBindingRequest(b.sbr, &b.sbr.Status)
```

Comment:
Please note that `updateStatusServiceBindingRequest` compares the first argument's `Status` field against the second argument's value before triggering the update; this change makes the comparison return always true, so the update won't be processed AFAICS.

Reply:
handled it here 
fcf6e52